tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+                }
+
+                cum_sums
+                    .par_iter_mut()
+                    .zip(luts.par_iter())
+                    .for_each(|(cum_sum_block, lut)| {
+                        self.key.apply_lookup_table_assign(cum_sum_block, lut);
+                    });
+
+                // Cum sums now contains the output carries
+                resolved_carries.append(&mut cum_sums);
+            }
+
+            resolved_carries
+        } else {
+            let lut_carry_propagation_sum =
+                self.key
+                    .generate_lookup_table_bivariate(|msb: u64, lsb: u64| -> u64 {
+                        if msb == 2 {
+                            1 // Remap Generate to 1
+                        } else if msb == 3 {
+                            // MSB propagates
+                            if lsb == 2 {
+                                1
+                            } else {
+                                lsb
+                            } // also remap here
+                        } else {
+                            msb
+                        }
+                    });
+            let sum_function = |block_borrow: &mut Ciphertext,
+                                previous_block_borrow: &Ciphertext| {
+                self.key.unchecked_apply_lookup_table_bivariate_assign(
+                    block_borrow,
+                    previous_block_borrow,
+                    &lut_carry_propagation_sum,
+                );
+            };
+            let mut resolved_carries =
+                self.compute_prefix_sum_hillis_steele(groupings_pgns, sum_function);
+            resolved_carries.insert(0, self.key.create_trivial(0));
+            resolved_carries
+        };
+
+        // Final step: adding resolved carries and cleaning result
+        let mut add_carries_and_cleanup = || {
+            let message_extract_lut = self
+                .key
+                .generate_lookup_table(|block| (block >> 1) % message_modulus);
+
+            prepared_blocks
+                .par_iter_mut()
+                .enumerate()
+                .for_each(|(i, block)| {
+                    let grouping_index = i / grouping_size;
+                    let borrow = &resolved_carries[grouping_index];
+                    crate::core_crypto::algorithms::lwe_ciphertext_add_assign(
+                        &mut block.ct,
+                        &borrow.ct,
+                    );
+
+                    self.key
+                        .apply_lookup_table_assign(block, &message_extract_lut)
+                });
+        };
+
+        match requested_flag {
+            OutputFlag::None => {
+                add_carries_and_cleanup();
+            }
+            OutputFlag::Overflow => {
+                let overflow_flag_lut = self.key.generate_lookup_table(|block| {
+                    let input_carry = (block >> 1) & 1;
+                    if input_carry == 1 {
+                        (block >> 3) & 1
```

Comment:
overflow encoding ?

Reply:
I put some intermediate variables to make it clearer